In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv(r"C:\Users\gjrah\Documents\Data Science\Data_Science\Statistics\Datasets\Modified_Z_Score\movie_revenues.csv")

In [3]:
df.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1   54000000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
2   21000000  [{"id": 18, "name": "Drama"}, {"id": 53, "name...   
3    1000000  [{"id": 80, "name": "Crime"}, {"id": 18, "name...   
4  100000000  [{"id": 18, "name": "Drama"}, {"id": 878, "nam...   

                         homepage     id  \
0     http://www.avatarmovie.com/  19995   
1  http://www.youmeanddupree.com/   1819   
2                             NaN   2575   
3                             NaN  26791   
4                             NaN    644   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 1253, "name": "roommate"}, {"id": 2038...                en   
2  [{"id": 246, "name": "dancing"}, {"id": 470, "...                en   
3                                                 []                en   
4  [{"id": 310, "name": "artificial intelligence"...                en   

                 original_title  \
0                        Avatar   
1            You, Me and Dupree   
2          The Tailor of Panama   
3                  Brigham City   
4  A.I. Artificial Intelligence   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  After standing in as best man for his longtime...   18.600367   
2  A British spy is banished to Panama after havi...    7.047975   
3  Wes Clayton is a lawman and a bishop in a Morm...    0.280083   
4  A robotic boy, the first programmed to love, D...   34.035114   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Universal Pictures", "id": 33}, {"n...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Main Street Movie Company", "id": 6...   
4  [{"name": "DreamWorks SKG", "id": 27}, {"name"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   12/10/2009  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...    7/14/2006   130431368   
2  [{"iso_3166_1": "IE", "name": "Ireland"}, {"is...    2/11/2001    28008462   
3  [{"iso_3166_1": "US", "name": "United States o...    3/30/2001      852206   
4  [{"iso_3166_1": "US", "name": "United States o...    6/29/2001   235926552   

   runtime                                   spoken_languages    status  \
0      162  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1      108           [{"iso_639_1": "en", "name": "English"}]  Released   
2      109  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
3      119           [{"iso_639_1": "en", "name": "English"}]  Released   
4      146           [{"iso_639_1": "en", "name": "English"}]  Released   

                                             tagline  \
0                        Enter the World of Pandora.   
1                   Two's company. Dupree's a crowd.   
2  In a place this treacherous, what a good spy n...   
3                                                NaN   
4  David is 11 years old. He weighs 60 pounds. He...   

                          title  vote_average  vote_count  
0                        Avatar           7.2       11800  
1            You, Me and Dupree           5.4         407  
2          The Tailor of Panama           6.2          92  
3                  Brigham City           7.3           4  
4  A.I. Artificial Intelligence           6.8        1974

- We can observe that the review and the budget are in 10^6

In [5]:
#Describe
df.describe()

budget             id  popularity       revenue     runtime  \
count  4.600000e+01      46.000000   46.000000  4.600000e+01   46.000000   
mean   4.549239e+07   57601.782609   40.342615  1.879289e+08  110.413043   
std    4.529615e+07   88002.107776   64.479614  4.551144e+08   17.994538   
min    4.500000e+05      59.000000    0.280083  8.522060e+05   77.000000   
25%    2.125000e+07    3156.750000   11.610217  2.866957e+07   96.250000   
50%    3.000000e+07   12506.500000   30.140234  8.381714e+07  108.000000   
75%    5.500000e+07   52522.500000   38.918290  1.382135e+08  121.750000   
max    2.370000e+08  298312.000000  418.708552  2.787965e+09  162.000000   

       vote_average    vote_count  
count     46.000000     46.000000  
mean       6.250000   1233.413043  
std        0.694342   2093.319773  
min        4.700000      4.000000  
25%        5.825000    251.500000  
50%        6.300000    733.000000  
75%        6.700000   1238.000000  
max        7.500000  11800.000000

#### Data Normalization
- Let's normalize the revenue data
- To normalize the data, I would use divide each data point using 10^6 since each data point is in 10^6

In [6]:
#Normalizing the revenue data
df['normalized_revenue'] = df['revenue'].apply(lambda x: x/1000000)

In [8]:
#2 random samples from the normalized data
df['normalized_revenue'].sample(2)

9    150.483765
2     28.008462
Name: normalized_revenue, dtype: float64

#### Explanation of the sample data
- We have calculated for 10^6, which means 10 lakhs
- We can observe that index 9 has 150.48 which means 15 cores in revenue
- Index 2 has 28.00 which means 28 cores of revenue

In [9]:
#Descrbing the data
df['normalized_revenue'].describe()

count      46.000000
mean      187.928898
std       455.114423
min         0.852206
25%        28.669569
50%        83.817142
75%       138.213502
max      2787.965087
Name: normalized_revenue, dtype: float64

In [14]:
_,mean,std,*_ = df['normalized_revenue'].describe()
print(f"mean:{mean} \nstandard deviation: {std}")

mean:187.92889841304347 
standard deviation: 455.1144234195408


In [15]:
#Function for z score
def get_z_score(value,mean,std):
    return (value-mean)/std

df['zscore'] = df.normalized_revenue.apply(lambda x: get_z_score(x,mean,std))

In [17]:
print(df['zscore'].head(3))

0    5.712929
1   -0.126336
2   -0.351385
Name: zscore, dtype: float64


In [19]:
#Function for Mean Absolute Distribution
def get_MAD(s):
    median = np.median(s)
    diff = abs(s - median)
    MAD = np.median(diff)
    return MAD

In [22]:
#Defining variables
MAD = get_MAD(df.normalized_revenue)
median = np.median(df.normalized_revenue)

#Modified zscore
def get_modified_zscore(value,median,MAD):
    return 0.675*(value-median)/MAD

In [29]:
#Calculating Modified Z-Score
df['mod_zscore'] = df.normalized_revenue.apply(lambda x: get_modified_zscore(x,median,MAD))

In [30]:
df.head(3)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1   54000000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
2   21000000  [{"id": 18, "name": "Drama"}, {"id": 53, "name...   

                         homepage     id  \
0     http://www.avatarmovie.com/  19995   
1  http://www.youmeanddupree.com/   1819   
2                             NaN   2575   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 1253, "name": "roommate"}, {"id": 2038...                en   
2  [{"id": 246, "name": "dancing"}, {"id": 470, "...                en   

         original_title                                           overview  \
0                Avatar  In the 22nd century, a paraplegic Marine is di...   
1    You, Me and Dupree  After standing in as best man for his longtime...   
2  The Tailor of Panama  A British spy is banished to Panama after havi...   

   popularity                               production_companies  ...  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   
1   18.600367  [{"name": "Universal Pictures", "id": 33}, {"n...  ...   
2    7.047975  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                                             tagline                 title  \
0                        Enter the World of Pandora.                Avatar   
1                   Two's company. Dupree's a crowd.    You, Me and Dupree   
2  In a place this treacherous, what a good spy n...  The Tailor of Panama   

  vote_average vote_count normalized_revenue    zscore  mod-zscore  mod_zscore  
0          7.2      11800        2787.965087  5.712929   32.363735   32.363735  
1          5.4        407         130.431368 -0.126336    0.557888    0.557888  
2          6.2         92          28.008462 -0.351385   -0.667928   -0.667928  

[3 rows x 24 columns]

In [34]:
df[df.mod_zscore > 3.5]

budget                                             genres  \
0   237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
7   150000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
12   25000000  [{"id": 53, "name": "Thriller"}, {"id": 28, "n...   

                         homepage      id  \
0     http://www.avatarmovie.com/   19995   
7   http://www.jurassicworld.com/  135397   
12                            NaN     619   

                                             keywords original_language  \
0   [{"id": 1463, "name": "culture clash"}, {"id":...                en   
7   [{"id": 1299, "name": "monster"}, {"id": 1718,...                en   
12  [{"id": 1156, "name": "sister sister relations...                en   

    original_title                                           overview  \
0           Avatar  In the 22nd century, a paraplegic Marine is di...   
7   Jurassic World  Twenty-two years after the events of Jurassic ...   
12   The Bodyguard  A former Secret Service agent grudgingly takes...   

    popularity                               production_companies  ...  \
0   150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   
7   418.708552  [{"name": "Universal Studios", "id": 13}, {"na...  ...   
12   26.576385  [{"name": "Tig Productions", "id": 335}, {"nam...  ...   

                                     spoken_languages    status  \
0   [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
7            [{"iso_639_1": "en", "name": "English"}]  Released   
12           [{"iso_639_1": "en", "name": "English"}]  Released   

                                              tagline           title  \
0                         Enter the World of Pandora.          Avatar   
7                                   The park is open.  Jurassic World   
12  Never let her out of your sight. Never let you...   The Bodyguard   

   vote_average vote_count normalized_revenue    zscore  mod-zscore  \
0           7.2      11800        2787.965087  5.712929   32.363735   
7           6.5       8662        1513.528810  2.912674   17.111050   
12          6.1        661         411.006740  0.490158    3.915865   

    mod_zscore  
0    32.363735  
7    17.111050  
12    3.915865  

[3 rows x 24 columns]